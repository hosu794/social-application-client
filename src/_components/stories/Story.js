import React, { useEffect } from "react";

import { storyActions, userActions } from "../../_actions";
import { useDispatch, useSelector } from "react-redux";

function Story(props) {
  const dispatch = useDispatch();
  const isLoading = useSelector((state) => state.stories.loading);
  const story = useSelector((state) => state.stories.currentStory);
  const loadedUser = useSelector((state) => state.user.loadedUser);
  const user = useSelector((state) => state.user.user);
  const loaded = !isLoading && loadedUser && story;

  useEffect(() => {
    dispatch(storyActions.getStoryById(props.match.params.id));
  }, []);

  return (
    <section class="section">
      {loaded ? (
        <LoadedStory
          title={story.title}
          description={story.description}
          key={story.id}
          creator={story.createdBy}
          loves={story.totalLoves}
          topic={story.topic}
          body={story.body}
          id={story.id}
          user={user}
        />
      ) : (
        "Loading..."
      )}
    </section>
  );
}

function LoadedStory(props) {
  const dispatch = useDispatch();

  const isUserLovedStory = useSelector((state) => state.user.isUserLovedStory);

  useEffect(() => {
    dispatch(
      userActions.checkLoveAvailability({
        storyId: props.id,
        userId: props.user.id,
      })
    );
  }, [props.loves]);

  function castStore(e) {
    e.preventDefault();
    dispatch(storyActions.castLove({ storyId: props.id }));
  }

  function unCastStore(e) {
    e.preventDefault();
    dispatch(storyActions.unCastLove({ storyId: props.id }));
  }

  return (
    <section class="section">
      <div class="container">
        <h1 className="title">{props.title}</h1>
        <h2 class="subtitle">{props.description}</h2>
        <p>{props.body}</p>
        <nav
          class="level"
          style={{
            marginTop: "1em",
          }}
        >
          <div class="level-item has-text-centered">
            <div>
              <p class="heading">Author</p>
              <p class="title">{props.creator.username}</p>
            </div>
          </div>
          <div class="level-item has-text-centered">
            <div>
              <p class="heading">Topic</p>
              <p class="title">{props.topic.title}</p>
            </div>
          </div>
          <div class="level-item has-text-centered">
            <div>
              <p class="heading">Loves</p>
              <p class="title">{props.loves}</p>
            </div>
          </div>
          <div class="level-item has-text-centered">
            <div>
              {isUserLovedStory ? (
                <button onClick={castStore} className="button is-white">
                  Like
                </button>
              ) : (
                <button onClick={unCastStore} className="button is-white">
                  Unlike
                </button>
              )}
            </div>
          </div>
        </nav>
      </div>
    </section>
  );
}

export default Story;
