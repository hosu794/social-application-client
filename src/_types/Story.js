import PropTypes from "prop-types";
import { Component } from "react";

export const IStory = (Component.PropTypes = {
  topic: PropTypes.object.isRequired,
  createdBy: PropTypes.object.isRequired,
  id: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  totalLoves: PropTypes.number.isRequired,
});
