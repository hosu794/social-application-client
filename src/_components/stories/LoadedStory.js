import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { storyActions, userActions } from "../../_actions";
import { useDispatch, useSelector } from "react-redux";
import { IUser, IStory } from "../../_types";

import CastStoreButton from "./CastStoreButton";

import UnCastStoreButton from "./UnCastStoreButton";

import DeleteStoryButton from "./DeleteStoryButton";
import { Link } from "react-router-dom";

function LoadedStory(props) {
  const dispatch = useDispatch();

  const editStoryUrl = `/stories/edit/${props.id}`;

  const isUserLovedStory = useSelector((state) => state.user.isUserLovedStory);
  const isLogged = useSelector((state) => state.authentication.loggedIn);
  const user = useSelector((state) => state.user.user);
  const isUserExist = user ? user.id : null;
  const isUserIdentificationIsEqualStoryCreator = isUserExist === props.userId;
  const currentStory = useSelector((state) => state.stories.currentStory);

  function deleteHandler() {
    dispatch(storyActions.deleteStory(props.id));
  }

  const unCastStore = async () => {
    await dispatch(storyActions.unCastLove({ storyId: props.id }));
  };

  const castStore = async () => {
    await dispatch(storyActions.castLove({ storyId: props.id }));
  };

  const checkLoveAvailabilityIfIsLogged = async () => {
    if (isLogged) {
      await dispatch(
        userActions.checkLoveAvailability({
          storyId: currentStory.id,
          userId: isUserExist,
        })
      );
    }
  };

  useEffect(() => {
    checkLoveAvailabilityIfIsLogged();
  }, [props.loves]);

  return (
    <section className="section">
      <div className="container">
        <h1 className="title">{props.title}</h1>
        <h2 className="subtitle">{props.description}</h2>
        <p dangerouslySetInnerHTML={{ __html: props.body }}></p>
        <nav
          className="level"
          style={{
            marginTop: "1em",
          }}
        >
          <div className="level-item has-text-centered">
            <div>
              <p className="heading">Author</p>
              <p className="title">{props.creator.username}</p>
            </div>
          </div>
          <div className="level-item has-text-centered">
            <div>
              <p className="heading">Topic</p>
              <p className="title">{props.topic.title}</p>
            </div>
          </div>
          <div className="level-item has-text-centered">
            <div>
              <p className="heading">Loves</p>
              <p className="title">{props.loves}</p>
            </div>
          </div>
          <div className="level-item has-text-centered">
            {isLogged ? (
              <div>
                {isUserLovedStory ? (
                  <CastStoreButton id={props.id} castStore={castStore} />
                ) : (
                  <UnCastStoreButton id={props.id} unCastStore={unCastStore} />
                )}
              </div>
            ) : null}
          </div>
          {isUserIdentificationIsEqualStoryCreator ? (
            <div class="level-item has-text-centered">
              <div>
                <DeleteStoryButton deleteHandler={deleteHandler} />
              </div>
            </div>
          ) : null}
          {isUserIdentificationIsEqualStoryCreator ? (
            <div class="level-item has-text-centered">
              <div>
                <Link to={editStoryUrl}>
                  <button className="button is-info">Edit</button>
                </Link>
              </div>
            </div>
          ) : null}
        </nav>
      </div>
    </section>
  );
}

LoadedStory.propTypes = {
  isUserLovedStory: PropTypes.bool,
  user: PropTypes.instanceOf(IUser),
  props: PropTypes.shape({
    title: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    body: PropTypes.string.isRequired,
    loves: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    creator: PropTypes.shape({
      username: PropTypes.string.isRequired,
    }),
  }),
};

export default LoadedStory;
