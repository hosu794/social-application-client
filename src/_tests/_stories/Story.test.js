import React from "react";

import configureMock from "redux-mock-store";
import { create } from "lodash";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import Story from "../../_components/stories/Story";
import { mount } from "enzyme";
import LoadedStory from "../../_components/stories/LoadedStory";
import { stories } from "../../_reducers/story.reducer";

let mockStore = configureMock([]);

let store = mockStore({
  comment: {
    content: [
      {
        id: 1,
        body: "Some body",
        createdBy: {
          id: 12,
          name: "Joe Doe",
          username: "exampleusername",
          downloadAvatar: "dsdasdasdasdasdadas.png",
        },
      },
    ],
  },
  authentication: {
    loggedIn: true,
  },
  stories: {
    currentStory: {
      id: 1,
      title: "Story Title",
      description: "Story description",
      totalLoves: 12,
      body: "<p>Simple Story</p>",
      createdBy: {
        id: 12,
        name: "Joe Doe",
        username: "exampleusername",
        downloadAvatar: "dsdasdasdasdasdadas.png",
      },
      topic: {
        id: 12,
        title: "Topic Example Title",
        description: "Description Example description",
        createdAt: "Dummy date",
        createdBy: {
          id: 12,
          name: "Joe Doe",
          username: "exampleusername",
          downloadAvatar: "dsdasdasdasdasdadas.png",
        },
      },
    },
    content: [
      {
        id: 1,
        title: "Story Title",
        description: "Story description",
        totalLoves: 12,
        body: "<p>Simple Story</p>",
        createdBy: {
          id: 12,
          name: "Joe Doe",
          username: "exampleusername",
          downloadAvatar: "dsdasdasdasdasdadas.png",
        },
        topic: {
          id: 12,
          title: "Topic Example Title",
          description: "Description Example description",
          createdAt: "Dummy date",
          createdBy: {
            id: 12,
            name: "Joe Doe",
            username: "exampleusername",
            downloadAvatar: "dsdasdasdasdasdadas.png",
          },
        },
      },
    ],
    size: 12,
    totalPages: 2,
    loading: false,
    error: false,
    casting: false,
    currentPage: 0,
    isLoading: false,
  },

  user: {
    user: {
      id: 12,
      name: "Joe Doe",
      username: "exampleusername",
      downloadAvatar: "dsdasdasdasdasdadas.png",
    },
    loadedUser: true,
    isLoading: false,
  },
});

let mockMatch = {
  params: {
    id: 12,
  },
};

describe("Test for the Story Component", () => {
  let TestingComponent = () => (
    <MemoryRouter>
      <Provider store={store}>
        <Story match={mockMatch} />
      </Provider>
    </MemoryRouter>
  );

  let wrapper = mount(<TestingComponent />);

  test("should component render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should component be defined", () => {
    expect(Story).toBeDefined();
  });

  test("should LoadedStory exists", () => {
    let element = wrapper.find(".section").first().text();

    const expected = "Loading...";

    expect(element).toEqual(expect.not.stringContaining(expected));
  });
});
