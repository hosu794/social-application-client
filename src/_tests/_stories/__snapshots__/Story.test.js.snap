// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the Story Component should component render correctly 1`] = `ReactWrapper {}`;
